### ISSUE TITLE
Proposal: telemetry event definitions

### ISSUE THREAD

**Issue by bryannaegele** [beam-telemetry/telemetry_registry#4]
## Background

Users currently rely on discovery and documentation of telemetry events in two manners: reading source code and documentation. To make use of events users must have an understanding of the following:
* event name
* available measurements
* available metadata

The general pattern in use today which we've iterated towards in Elixir is:

* `[:my_event, :name]` - general description of what the event is
  * Measurement: `%{map_of: pseudo_type_spec}`
  * Metadata: `%{map_of: pseudo_type_spec}`

This has worked fairly well for most users I've interacted with, so we should build on this as a starting point to support documentation and discovery.

## Proposal

Telemetry module event attributes should support an event definition map, in addition to a plain event, such as `[my_app, event, stop]` to provide a standard for documentation which can be leveraged by tools like `ex_doc`

### Example

```erlang
-telemetry_event([my_app, some, event]).
-telemetry_event(#{
    event => [my_app, event, stop],
    description => "A description of the event",
    measurement => "`#{duration => non_neg_integer()}`",
    metadata => "`#{status => 200 | 400 | 404, method => string()}`"
}).
```

Plain events would be stored with empty string defaults for the other properties of a definition.

## Discussion

While it would be nifty to use actual typespecs for the measurement and metadata, I'm unfamiliar with any way to accomplish that, especially cross-language. If there isn't a way to leverage typespecs, a plain string with the user using a single-line code block of pseduo-typespec feels sufficient. The primary use case for the measurement and metadata information is to inform users of what is available.

### Status: Accepted

**Comment by tsloughter** [comment]
So Elixir lets users add module attributes like these as well?

**Comment by tsloughter** [comment]
Another option that would allow typespecs would be to actually use a type spec::

```
-type telemetry_events() :: [#{
    event => [my_app, event, stop],
    description => "A description of the event",
    measurement => #{duration => non_neg_integer()},
    metadata => #{status => 200 | 400 | 404, method: string()}
}].
```

**Comment by tsloughter** [comment]
Wait, scratch that, you can't have a type like `[my_app, event, stop]` since it doesn't understand that.

**Comment by bryannaegele** [comment]
Yep. Atom literal lists and string literals are not supported in types, so I don't think there is an immediate path there.

And yes, Elixir users would do the same. They have to set aggregate to true, as well.

```elixir
Module.register_attribute(__MODULE__, :telemetry_event, accumulate: true, persist: true)
@telemetry_event [:my_app, :some, :event]
@telemetry_event %{
    event: [:my_app, :event, :stop],
    description: "A description of the event",
    measurement: "`%{duration: non_neg_integer()}`",
    metadata: "`%{status: 200 | 400 | 404, method: string()}`"
})
```

To improve user experience in both languages, we can add span macros to automatically add the attributes for auto-discovery, but that's outside the scope of this particular discussion. A using macro to add the attr registration would be nice, as well as the other one mentioned for spans. But I have no idea how we'd package up those macros. Maybe we end up with an Elixir specific library that exposes conveniences, whether in the org or outside.

**Comment by bryannaegele** [comment]
I would like to discuss this proposal and a path towards inclusion to `telemetry` itself at the WG meeting this week, so any additional feedback is appreciated.

/cc @josevalim @gregmefford @arkgil

**Comment by bryannaegele** [comment]
Resolved with #5
